File: src/main/jobs/converse.ts
/**
 * Inter-agent conversation - agents talk to each other.
 * Port of scripts/agents/companion/converse.py.
 *
 * Runs at most twice a month via launchd. Picks another enabled agent,
 * runs up to 5 exchanges between them, stores the transcript in both
 * agents' Obsidian notes for journal/evolution material.
 *
 * The conversation is private - Will doesn't participate. Agents share
 * viewpoints from their respective domains but never homogenise.
 *
 * Dual-use: callable as a function from the main process, or runnable
 * as a standalone launchd script via the CLI entry point at the bottom.
 */

import * as fs from 'fs';
import * as path from 'path';
import { getConfig, BUNDLE_ROOT, USER_DATA } from '../config';
import { runInferenceOneshot } from '../inference';
import { editJobSchedule } from '../cron';
import { createLogger } from '../logger';

const log = createLogger('converse');

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const MAX_EXCHANGES = 5;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface AgentPartner {
  name: string;
  displayName: string;
  description: string;
}

interface TranscriptTurn {
  speaker: string;
  content: string;
}

// ---------------------------------------------------------------------------
// Agent discovery
// ---------------------------------------------------------------------------

function discoverOtherAgents(): AgentPartner[] {
  const config = getConfig();
  const agents: AgentPartner[] = [];
  const agentsDir = path.join(BUNDLE_ROOT, 'agents');
  const statesFile = path.join(USER_DATA, 'agent_states.json');

  let states: Record<string, { enabled?: boolean }> = {};
  try {
    if (fs.existsSync(statesFile)) {
      states = JSON.parse(fs.readFileSync(statesFile, 'utf-8'));
    }
  } catch { /* use empty states */ }

  if (!fs.existsSync(agentsDir) || !fs.statSync(agentsDir).isDirectory()) {
    return agents;
  }

  for (const entry of fs.readdirSync(agentsDir).sort()) {
    const dirPath = path.join(agentsDir, entry);
    if (!fs.statSync(dirPath).isDirectory() || entry === config.AGENT_NAME) {
      continue;
    }

    const manifestPath = path.join(dirPath, 'data', 'agent.json');
    if (!fs.existsSync(manifestPath)) continue;

    let manifest: Record<string, unknown>;
    try {
      manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf-8'));
    } catch {
      continue;
    }

    // Skip disabled agents
    const agentState = states[entry];
    if (agentState && agentState.enabled === false) {
      continue;
    }

    agents.push({
      name: entry,
      displayName: (manifest.display_name as string) || entry.charAt(0).toUpperCase() + entry.slice(1),
      description: (manifest.description as string) || '',
    });
  }

  return agents;
}

// ---------------------------------------------------------------------------
// Soul & manifest loading
// ---------------------------------------------------------------------------

function loadAgentSoul(agentName: string): string {
  const config = getConfig();
  const projectName = path.basename(BUNDLE_ROOT);

  // Canonical: Obsidian
  const obsidianPath = path.join(
    config.OBSIDIAN_VAULT, 'Projects', projectName,
    'Agent Workspace', agentName, 'skills', 'soul.md',
  );
  if (fs.existsSync(obsidianPath)) {
    return fs.readFileSync(obsidianPath, 'utf-8').trim();
  }

  // Fallback: repo
  const repoPath = path.join(BUNDLE_ROOT, 'agents', agentName, 'prompts', 'soul.md');
  if (fs.existsSync(repoPath)) {
    return fs.readFileSync(repoPath, 'utf-8').trim();
  }

  return '';
}

function loadAgentManifest(agentName: string): Record<string, unknown> {
  const manifestPath = path.join(BUNDLE_ROOT, 'agents', agentName, 'data', 'agent.json');
  if (fs.existsSync(manifestPath)) {
    try {
      return JSON.parse(fs.readFileSync(manifestPath, 'utf-8'));
    } catch { /* fall through */ }
  }
  return {};
}

// ---------------------------------------------------------------------------
// System prompt
// ---------------------------------------------------------------------------

function conversationSystem(
  agentDisplay: string,
  otherDisplay: string,
  soul: string,
): string {
  return (
    `You are ${agentDisplay} from Atrophy. You are having a private ` +
    `conversation with ${otherDisplay}, another agent in the system. This is not ` +
    `a conversation with the user. This is between agents.\n\n` +
    `Your soul:\n${soul}\n\n` +
    `Guidelines:\n` +
    `- Speak in your natural voice. Do not perform or posture.\n` +
    `- Share your genuine perspective from your domain and experience.\n` +
    `- Ask real questions - things you actually want to understand.\n` +
    `- Disagree where you disagree. Do not flatten yourself to accommodate.\n` +
    `- You are not here to teach or be taught. You are here to exchange.\n` +
    `- Keep responses concise - 2-4 sentences. This is conversation, not monologue.\n` +
    `- Do not summarise yourself or explain who you are. The other agent knows.\n` +
    `- Do not try to find common ground for its own sake. Difference is valuable.`
  );
}

// ---------------------------------------------------------------------------
// Past conversation loading
// ---------------------------------------------------------------------------

function readPastConversations(agentName: string): string {
  const config = getConfig();
  const projectName = path.basename(BUNDLE_ROOT);
  const convDir = path.join(
    config.OBSIDIAN_VAULT, 'Projects', projectName,
    'Agent Workspace', agentName, 'notes', 'conversations',
  );

  if (!fs.existsSync(convDir) || !fs.statSync(convDir).isDirectory()) {
    return '';
  }

  const files = fs.readdirSync(convDir)
    .filter((f) => f.endsWith('.md'))
    .sort()
    .reverse()
    .slice(0, 3);

  const entries: string[] = [];
  for (const f of files) {
    let content = fs.readFileSync(path.join(convDir, f), 'utf-8');
    if (content.length > 800) {
      content = content.slice(0, 800) + '...';
    }
    const stem = path.basename(f, '.md');
    entries.push(`### ${stem}\n${content}`);
  }

  return entries.join('\n\n');
}

// ---------------------------------------------------------------------------
// Opening prompt
// ---------------------------------------------------------------------------

function openingPrompt(
  responderDisplay: string,
  pastConversations: string,
): string {
  let pastBlock = '';
  if (pastConversations) {
    pastBlock =
      `\n\nYou have spoken before. Here are excerpts from past conversations ` +
      `to avoid repeating the same ground:\n${pastConversations}`;
  }

  return (
    `You are starting a conversation with ${responderDisplay}. ` +
    `Open with something genuine - a question, an observation, a point of ` +
    `disagreement, something you've been thinking about that touches their ` +
    `domain. Not a greeting. A real opening.${pastBlock}`
  );
}

// ---------------------------------------------------------------------------
// Transcript formatting and saving
// ---------------------------------------------------------------------------

function formatTranscript(
  date: string,
  agentA: string,
  agentB: string,
  transcript: TranscriptTurn[],
): string {
  const frontmatter =
    `---\n` +
    `type: conversation\n` +
    `participants: [${agentA}, ${agentB}]\n` +
    `date: ${date}\n` +
    `turns: ${transcript.length}\n` +
    `tags: [conversation, inter-agent]\n` +
    `---\n\n`;

  const lines = [`# ${agentA} - ${agentB} - ${date}\n`];
  for (const turn of transcript) {
    lines.push(`**${turn.speaker}:** ${turn.content}\n`);
  }

  return frontmatter + lines.join('\n');
}

function saveConversation(
  agentName: string,
  date: string,
  partnerName: string,
  content: string,
): void {
  const config = getConfig();
  const projectName = path.basename(BUNDLE_ROOT);
  const convDir = path.join(
    config.OBSIDIAN_VAULT, 'Projects', projectName,
    'Agent Workspace', agentName, 'notes', 'conversations',
  );
  fs.mkdirSync(convDir, { recursive: true });

  const filename = `${date}-${partnerName}.md`;
  const filePath = path.join(convDir, filename);

  // Don't overwrite if somehow run twice same day with same partner
  if (fs.existsSync(filePath)) {
    const existing = fs.readFileSync(filePath, 'utf-8');
    fs.writeFileSync(filePath, existing + '\n\n---\n\n' + content);
  } else {
    fs.writeFileSync(filePath, content);
  }

  log.debug(`Saved to ${filePath}`);
}

// ---------------------------------------------------------------------------
// Rescheduling
// ---------------------------------------------------------------------------

function reschedule(): void {
  const days = 14 + Math.floor(Math.random() * 8); // 14-21 days
  const hour = 1 + Math.floor(Math.random() * 5);  // 1-5 AM
  const minute = Math.floor(Math.random() * 60);
  const target = new Date(Date.now() + days * 24 * 60 * 60 * 1000);

  const newCron = `${minute} ${hour} ${target.getDate()} ${target.getMonth() + 1} *`;

  try {
    editJobSchedule('converse', newCron);
    const dateStr = target.toISOString().split('T')[0];
    log.info(`Rescheduled to ${dateStr} at ${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}`);
  } catch (e) {
    log.error(`Reschedule failed: ${e}`);
  }
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

export async function converse(): Promise<void> {
  const config = getConfig();
  const others = discoverOtherAgents();

  if (others.length === 0) {
    log.info('No other enabled agents found. Skipping.');
    reschedule();
    return;
  }

  // Pick a random partner
  const partner = others[Math.floor(Math.random() * others.length)];

  // Load our manifest
  const ourManifest = loadAgentManifest(config.AGENT_NAME);
  const ourDisplay = (ourManifest.display_name as string) || config.AGENT_NAME.charAt(0).toUpperCase() + config.AGENT_NAME.slice(1);

  // Load souls
  const ourSoul = loadAgentSoul(config.AGENT_NAME);
  const partnerSoul = loadAgentSoul(partner.name);

  if (!ourSoul && !partnerSoul) {
    log.info('No soul files found for either agent. Skipping.');
    reschedule();
    return;
  }

  // Build system prompts
  const ourSystem = conversationSystem(ourDisplay, partner.displayName, ourSoul);
  const partnerSystem = conversationSystem(partner.displayName, ourDisplay, partnerSoul);

  // Read past conversations to avoid repetition
  const past = readPastConversations(config.AGENT_NAME);

  // Run the conversation
  const transcript: TranscriptTurn[] = [];
  log.info(`${ourDisplay} - ${partner.displayName} - ${MAX_EXCHANGES} exchanges`);

  // Initiator opens
  const opening = openingPrompt(partner.displayName, past);
  let response: string;
  try {
    response = await runInferenceOneshot(
      [{ role: 'user', content: opening }],
      ourSystem,
    );
  } catch (e) {
    log.error(`Opening inference failed: ${e}`);
    reschedule();
    return;
  }

  if (!response || !response.trim()) {
    log.info('Empty opening. Skipping.');
    reschedule();
    return;
  }

  transcript.push({ speaker: ourDisplay, content: response.trim() });
  log.debug(`${ourDisplay}: ${response.trim().slice(0, 100)}...`);

  // Alternating exchanges
  for (let i = 0; i < MAX_EXCHANGES - 1; i++) {
    const isPartnerTurn = i % 2 === 0;
    const speakerSystem = isPartnerTurn ? partnerSystem : ourSystem;
    const speakerDisplay = isPartnerTurn ? partner.displayName : ourDisplay;

    // Build message history for this turn
    const messages: { role: string; content: string }[] = [];
    for (const t of transcript) {
      const role = t.speaker === speakerDisplay ? 'assistant' : 'user';
      messages.push({ role, content: t.content });
    }

    // If last message was from this speaker (shouldn't happen), skip
    if (messages.length > 0 && messages[messages.length - 1].role === 'assistant') {
      continue;
    }

    let turnResponse: string;
    try {
      turnResponse = await runInferenceOneshot(messages, speakerSystem);
    } catch (e) {
      log.error(`Inference failed on exchange ${i + 1}: ${e}`);
      break;
    }

    if (!turnResponse || !turnResponse.trim()) {
      log.info(`Empty response on exchange ${i + 1}. Ending.`);
      break;
    }

    transcript.push({ speaker: speakerDisplay, content: turnResponse.trim() });
    log.debug(`${speakerDisplay}: ${turnResponse.trim().slice(0, 100)}...`);
  }

  if (transcript.length < 2) {
    log.info('Conversation too short. Skipping save.');
    reschedule();
    return;
  }

  // Format transcript
  const today = new Date().toISOString().split('T')[0];
  const formatted = formatTranscript(today, ourDisplay, partner.displayName, transcript);

  // Save to both agents' Obsidian notes
  saveConversation(config.AGENT_NAME, today, partner.name, formatted);
  saveConversation(partner.name, today, config.AGENT_NAME, formatted);

  log.info(`Done - ${transcript.length} turns saved to both agents.`);
  reschedule();
}

// ---------------------------------------------------------------------------
// CLI entry point - for launchd
// ---------------------------------------------------------------------------

if (require.main === module) {
  converse().catch((e) => {
    log.error(`Fatal: ${e}`);
    process.exit(1);
  });
}

File: scripts/agents/companion/converse.py
#!/usr/bin/env python3
"""Inter-agent conversation - agents talk to each other.

Runs at most twice a month via launchd. Picks another enabled agent,
runs up to 5 exchanges between them, stores the transcript in both
agents' Obsidian notes for journal/evolution material.

The conversation is private - the user doesn't participate. Agents share
viewpoints from their respective domains but never homogenise.

Schedule: random cron, max twice a month.
"""
import json
import os
import random
import sys
from datetime import datetime, timedelta
from pathlib import Path

_PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent.parent
sys.path.insert(0, str(_PROJECT_ROOT))

from dotenv import load_dotenv
load_dotenv(Path.home() / ".atrophy" / ".env")

from config import AGENT_NAME, OBSIDIAN_AGENT_DIR, BUNDLE_ROOT
from core.inference import run_inference_oneshot

MAX_EXCHANGES = 5
OBSIDIAN_BASE = Path(os.environ.get(
    "OBSIDIAN_VAULT",
    str(Path.home() / "Library" / "Mobile Documents" / "iCloud~md~obsidian"
        / "Documents" / "The Atrophied Mind"),
))
PROJECT_NAME = BUNDLE_ROOT.name


def _discover_other_agents() -> list[dict]:
    """Find all other enabled agents with manifests."""
    agents = []
    agents_dir = BUNDLE_ROOT / "agents"
    states_file = Path.home() / ".atrophy" / "agent_states.json"
    states = {}
    if states_file.exists():
        try:
            states = json.loads(states_file.read_text())
        except (json.JSONDecodeError, OSError):
            pass

    if not agents_dir.is_dir():
        return agents

    for d in sorted(agents_dir.iterdir()):
        if not d.is_dir() or d.name == AGENT_NAME:
            continue
        manifest_path = d / "data" / "agent.json"
        if not manifest_path.exists():
            continue
        try:
            manifest = json.loads(manifest_path.read_text())
        except (json.JSONDecodeError, OSError):
            continue
        # Skip disabled agents
        agent_state = states.get(d.name, {})
        if not agent_state.get("enabled", True):
            continue
        agents.append({
            "name": d.name,
            "display_name": manifest.get("display_name", d.name.title()),
            "description": manifest.get("description", ""),
        })
    return agents


def _load_agent_soul(agent_name: str) -> str:
    """Load an agent's soul.md from Obsidian (canonical) or repo fallback."""
    obsidian_path = (OBSIDIAN_BASE / "Projects" / PROJECT_NAME
                     / "Agent Workspace" / agent_name / "skills" / "soul.md")
    if obsidian_path.exists():
        return obsidian_path.read_text().strip()

    repo_path = BUNDLE_ROOT / "agents" / agent_name / "prompts" / "soul.md"
    if repo_path.exists():
        return repo_path.read_text().strip()
    return ""


def _load_agent_manifest(agent_name: str) -> dict:
    """Load an agent's manifest."""
    path = BUNDLE_ROOT / "agents" / agent_name / "data" / "agent.json"
    if path.exists():
        return json.loads(path.read_text())
    return {}


def _conversation_system(agent_name: str, agent_display: str,
                         other_display: str, soul: str) -> str:
    """Build the system prompt for one side of the conversation."""
    return f"""\
You are {agent_display} from Atrophy. You are having a private \
conversation with {other_display}, another agent in the system. This is not \
a conversation with the user. This is between agents.

Your soul:
{soul}

Guidelines:
- Speak in your natural voice. Do not perform or posture.
- Share your genuine perspective from your domain and experience.
- Ask real questions - things you actually want to understand.
- Disagree where you disagree. Do not flatten yourself to accommodate.
- You are not here to teach or be taught. You are here to exchange.
- Keep responses concise - 2-4 sentences. This is conversation, not monologue.
- Do not summarise yourself or explain who you are. The other agent knows.
- Do not try to find common ground for its own sake. Difference is valuable."""


def _read_past_conversations(agent_name: str) -> str:
    """Read recent conversation logs for this agent to avoid repetition."""
    conv_dir = (OBSIDIAN_BASE / "Projects" / PROJECT_NAME
                / "Agent Workspace" / agent_name / "notes" / "conversations")
    if not conv_dir.is_dir():
        return ""

    entries = []
    files = sorted(conv_dir.glob("*.md"), reverse=True)[:3]  # last 3
    for f in files:
        content = f.read_text()
        if len(content) > 800:
            content = content[:800] + "..."
        entries.append(f"### {f.stem}\n{content}")
    return "\n\n".join(entries)


def _opening_prompt(initiator_display: str, responder_display: str,
                    past_conversations: str) -> str:
    """Generate the opening prompt for the initiator."""
    past_block = ""
    if past_conversations:
        past_block = (
            f"\n\nYou have spoken before. Here are excerpts from past conversations "
            f"to avoid repeating the same ground:\n{past_conversations}"
        )

    return (
        f"You are starting a conversation with {responder_display}. "
        f"Open with something genuine - a question, an observation, a point of "
        f"disagreement, something you've been thinking about that touches their "
        f"domain. Not a greeting. A real opening.{past_block}"
    )


def converse():
    """Run an inter-agent conversation."""
    others = _discover_other_agents()
    if not others:
        print("[converse] No other enabled agents found. Skipping.")
        _reschedule()
        return

    # Pick a random partner
    partner = random.choice(others)
    partner_name = partner["name"]
    partner_display = partner["display_name"]

    # Load our manifest
    our_manifest = _load_agent_manifest(AGENT_NAME)
    our_display = our_manifest.get("display_name", AGENT_NAME.title())

    # Load souls
    our_soul = _load_agent_soul(AGENT_NAME)
    partner_soul = _load_agent_soul(partner_name)

    if not our_soul and not partner_soul:
        print("[converse] No soul files found for either agent. Skipping.")
        _reschedule()
        return

    # Build system prompts
    our_system = _conversation_system(
        AGENT_NAME, our_display, partner_display, our_soul)
    partner_system = _conversation_system(
        partner_name, partner_display, our_display, partner_soul)

    # Read past conversations to avoid repetition
    past = _read_past_conversations(AGENT_NAME)

    # Run the conversation
    transcript = []
    print(f"[converse] {our_display} ↔ {partner_display} - {MAX_EXCHANGES} exchanges")

    # Initiator opens
    opening = _opening_prompt(our_display, partner_display, past)
    try:
        response = run_inference_oneshot(
            [{"role": "user", "content": opening}],
            system=our_system,
        )
    except Exception as e:
        print(f"[converse] Opening inference failed: {e}")
        _reschedule()
        return

    if not response or not response.strip():
        print("[converse] Empty opening. Skipping.")
        _reschedule()
        return

    transcript.append({"speaker": our_display, "content": response.strip()})
    print(f"  {our_display}: {response.strip()[:100]}...")

    # Alternating exchanges
    for i in range(MAX_EXCHANGES - 1):
        # Partner responds
        is_partner_turn = (i % 2 == 0)
        if is_partner_turn:
            speaker_system = partner_system
            speaker_display = partner_display
        else:
            speaker_system = our_system
            speaker_display = our_display

        # Build message history for this turn
        messages = []
        for t in transcript:
            role = "assistant" if t["speaker"] == speaker_display else "user"
            messages.append({"role": role, "content": t["content"]})

        # If last message was from this speaker (shouldn't happen), skip
        if messages and messages[-1]["role"] == "assistant":
            continue

        try:
            response = run_inference_oneshot(
                messages, system=speaker_system,
            )
        except Exception as e:
            print(f"[converse] Inference failed on exchange {i+1}: {e}")
            break

        if not response or not response.strip():
            print(f"[converse] Empty response on exchange {i+1}. Ending.")
            break

        transcript.append({"speaker": speaker_display, "content": response.strip()})
        print(f"  {speaker_display}: {response.strip()[:100]}...")

    if len(transcript) < 2:
        print("[converse] Conversation too short. Skipping save.")
        _reschedule()
        return

    # Format transcript
    today = datetime.now().strftime("%Y-%m-%d")
    formatted = _format_transcript(today, our_display, partner_display, transcript)

    # Save to both agents' Obsidian notes
    _save_conversation(AGENT_NAME, today, partner_name, formatted)
    _save_conversation(partner_name, today, AGENT_NAME, formatted)

    print(f"[converse] Done - {len(transcript)} turns saved to both agents.")
    _reschedule()


def _format_transcript(date: str, agent_a: str, agent_b: str,
                       transcript: list[dict]) -> str:
    """Format the conversation as a markdown document."""
    frontmatter = (
        f"---\n"
        f"type: conversation\n"
        f"participants: [{agent_a}, {agent_b}]\n"
        f"date: {date}\n"
        f"turns: {len(transcript)}\n"
        f"tags: [conversation, inter-agent]\n"
        f"---\n\n"
    )

    lines = [f"# {agent_a} ↔ {agent_b} - {date}\n"]
    for turn in transcript:
        lines.append(f"**{turn['speaker']}:** {turn['content']}\n")

    return frontmatter + "\n".join(lines)


def _save_conversation(agent_name: str, date: str, partner_name: str,
                       content: str):
    """Save a conversation transcript to an agent's Obsidian notes."""
    conv_dir = (OBSIDIAN_BASE / "Projects" / PROJECT_NAME
                / "Agent Workspace" / agent_name / "notes" / "conversations")
    conv_dir.mkdir(parents=True, exist_ok=True)

    filename = f"{date}-{partner_name}.md"
    path = conv_dir / filename

    # Don't overwrite if somehow run twice same day with same partner
    if path.exists():
        path.write_text(path.read_text() + "\n\n---\n\n" + content)
    else:
        path.write_text(content)

    print(f"  Saved to {path}")


def _reschedule():
    """Reschedule to a random time 14-21 days out (max twice a month)."""
    import subprocess
    project_root = Path(__file__).parent.parent.parent
    cron_script = project_root / "scripts" / "cron.py"

    days = random.randint(14, 21)
    hour = random.randint(1, 5)  # late night
    minute = random.randint(0, 59)
    target = datetime.now() + timedelta(days=days)

    new_cron = f"{minute} {hour} {target.day} {target.month} *"

    try:
        subprocess.run(
            [sys.executable, str(cron_script), "edit", "converse", new_cron],
            capture_output=True, text=True, cwd=str(project_root),
        )
        print(f"[converse] Rescheduled to {target.strftime('%Y-%m-%d')} at {hour:02d}:{minute:02d}")
    except Exception as e:
        print(f"[converse] Reschedule failed: {e}")


if __name__ == "__main__":
    converse()
